Lib/zipimport.py

```diff
@@ -651,9 +671,25 @@ def _get_mtime_of_source(self, path):
         # with an embedded pyc time stamp
         time = toc_entry[5]
         date = toc_entry[6]
-        return _parse_dostime(date, time)
+        uncompressed_size = toc_entry[3]
+        return _parse_dostime(date, time), uncompressed_size
+    except (KeyError, IndexError, TypeError):
+        return 0, 0
+
+
+# Given a path to a .pyc file in the archive, return the
+# contents of the matching .py file, or None if no source
+# is available.
+def _get_pyc_source(self, path):
+    try:
+        # strip 'c' or 'o' from *.py[co]
+        assert path[-1:] in ('c', 'o')
+        path = path[:-1]
+        toc_entry = self._files[path]
+        return _get_data(self.archive, toc_entry)
     except (KeyError, IndexError, TypeError):
```

Comment:
In what circumstances IndexError and TypeError are raised?

Reply:
I copied this from `_get_mtime_of_source()` without really thinking.  I'm guessing only `KeyError` is relevant in this case (raised by `self._files[path]`).